userns/userns.go

```diff
@@ -29,27 +31,44 @@ import (
 )
 
 func init() {
-	err := WriteUsermap(map[uint32]uint32{
+	err := WriteMap("/proc/self/uid_map", map[uint32]uint32{
 		OriginalUID(): 0,
 	})
 	if err != nil {
-		panic(fmt.Errorf("failed to create user mapping: %v", err).Error())
+		panic(fmt.Errorf("failed to create uid mapping: %v", err).Error())
+	}
+
+	// write deny in setgroups to disable setgroup(2) and enable writing to gid_map
+	err = os.WriteFile("/proc/self/setgroups", []byte("deny\n"), 0o644)
+	if err != nil {
+		panic(fmt.Errorf("failed to deny setgroups: %v", err).Error())
+	}
+
+	err = WriteMap("/proc/self/gid_map", map[uint32]uint32{
+		OriginalGID(): 0,
+	})
+	if err != nil {
+		panic(fmt.Errorf("failed to create gid mapping: %v", err).Error())
 	}
 }
 
 func OriginalUID() uint32 {
 	return uint32(C.originalUid)
 }
 
-// WriteUsermap builds a map of Host UID -> Namespace UID.
+func OriginalGID() uint32 {
+	return uint32(C.originalGid)
+}
+
+// WriteMap builds a map of Host UID/GID -> Namespace UID/GID
 // Example:
 //
-//	WriteUsermap(map[uint32]uint32{userns.OriginalUID: 0, 1234: 1234})
-func WriteUsermap(mapping map[uint32]uint32) error {
+//	WriteMap(map[uint32]uint32{userns.OriginalUID: 0, 1234: 1234})
+func WriteMap(path string, mapping map[uint32]uint32) error {
```

Comment:
nit: I would keep `WriteUsermap` as a public api and add `WriteGroupMap`. They can share an internal `writeMap` function.

Reply:
Sure, done.
I have also moved the deny of setgroups inside the WriteGroupMap